### PR TITLE
Coverage pyfepdf

### PR DESCRIPTION
**Pull request by chazuttu**
## Summary

This PR adds unit tests for the script **pyfepdf.py** and also increase the coverage for this script.
For the above purpose, some additional files are also modified. 

- added unit tests for pyfepdf.py
- increased coverage to 80%  on python 3

## Checklist

- [x] Classes, Variables, function and methods logic  ok
- [x] Comments written explaining what the code does
- [x] All python code is PEP8 compliant (run black .)
- [x] No lint issues (run flake8)
- [x] Test coverage with pytest implemented
- [x] Reviewers assigned (at least 1 mentor)

## Manual test evidence
```
============================ test session starts ======================
platform win32 -- Python 3.9.6, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: C:\User\Week_7_GSOC\pyafipws, configfile: pytest.ini
plugins: cov-2.12.1, freezegun-0.4.2, html-3.1.1, metadata-1.11.0, vcr-1.0.2
collected 14 items

tests\test_pyfepdf.py ..............                                                                                                                                                               [100%]


----------- coverage: platform win32, python 3.9.6-final-0 -----------
Name                                                                              Stmts   Miss  Cover
-----------------------------------------------------------------------------------------------------
test3\Lib\site-packages\PyAfipWs-3.9.b_2500_-py3.9.egg\pyafipws\__init__.py           4      0   100%
test3\Lib\site-packages\PyAfipWs-3.9.b_2500_-py3.9.egg\pyafipws\pyfepdf.py          802    162    80%

======================= 14 passed, 2 warnings in 5.86s ============================
```

### Python 2
pytest --cov-config=.coveragerc --cov=pyafipws tests/test_pyfepdf.py
```
============================= test session starts =============================
platform win32 -- Python 2.7.18, pytest-4.6.11, py-1.10.0, pluggy-0.13.1
rootdir: C:\UsersDesktop\Week_7_GSOC\pyafipws, inifile: pytest.ini
plugins: cov-2.12.1, freezegun-0.4.2, html-1.22.1, metadata-1.11.0, vcr-1.0.2
collected 14 items

tests\test_pyfepdf.py ..............                                     [100%]



---------- coverage: platform win32, python 2.7.18-final-0 -----------
Name                                                                           Stmts   Miss  Cover
--------------------------------------------------------------------------------------------------
test2\Lib\site-packages\PyAfipWs-2.7.2500-py2.7.egg\pyafipws\__init__.py           4      0   100%
test2\Lib\site-packages\PyAfipWs-2.7.2500-py2.7.egg\pyafipws\pyfepdf.py          801    286    64%


==================== 14 passed, 1 warnings in 3.67 seconds ====================
```

### CODE REVIEW

File: .coveragerc [hunk 1]
```diff
@@ -6,7 +6,7 @@ omit =
         *nsis*
         */padron.py
         *pyemail*
-        *pyfepdf*
+        *formatos*
```
Comment (reingart) [inline .coveragerc:9]: formatos should be included, please create a new ticket for a new PR
(no need to add coverage in this one)

File: formatos/formato_json.py [hunk 2]
```diff
@@ -46,6 +46,6 @@ def escribir(filas, fn="salida.json"):
         jsonfile,
         sort_keys=True,
         indent=4,
-        encoding="utf-8",
+        # encoding="utf-8",
```
Comment (reingart) [inline formatos/formato_json.py]: In python 2 the encoding is needed, does this keep working?
You could use **kwargs and only pass encoding in python2

File: formatos/formato_txt.py [hunk 3]
```diff
@@ -257,7 +259,7 @@ def escribir_linea_txt(dic, formato):
 
 def leer(fn="entrada.txt"):
     "Analiza un archivo TXT y devuelve un diccionario"
-    f_entrada = open(fn, "r")
+    f_entrada = open(fn, "rb")
```
Comment (reingart) [inline formatos/formato_txt.py:262]: Again, this could fail if the PDF invoice has non-English characters
This functions should have individual Unit Test for this corner cases (for example: test files with accents áñ etc.)
Create another ticket to address this separately

File: pyfepdf.py [hunk 4]
```diff
@@ -32,7 +32,7 @@
 
 DEBUG = False
 HOMO = False
-CONFIG_FILE = "rece.ini"
+CONFIG_FILE = "conf/rece.ini"
```
Comment (reingart) [inline pyfepdf.py]: - [ ] Please do not change the default config file path
In real installation this avoids to overwrite the user file with the default
If you need, you could copy the file from conf/ to the base directory  in the GitHub Action workflow step

File: pyfepdf.py [hunk 5]
```diff
@@ -2037,7 +2037,7 @@ def main():
             elif "--json" in sys.argv:
                 from .formatos import formato_json
 
-                archivo = conf_fact.get("entrada", "entrada.txt")
+                archivo = "salida.json"
```
Comment (reingart) [inline pyfepdf.py]: - [ ] Please do not change the code semantics when testing
Here the file name should come from the configuration
You could use pytest mock to alter certain parameter in the test

File: tests/test_pyfepdf.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+def test_generar_qr():
+    fepdf.CUIT = "30000000007"
+    assert fepdf.GenerarQR()
+
+def test_main_prueba():
+    sys.argv = []
+    sys.argv.append("--prueba")
+    sys.argv.append("--debug")
+    main()
```
Comment (reingart) [inline tests/test_pyfepdf.py:284]: Please add a new ticket to compare the invoice generated
For example you could add the factura.pdf created in your machine and then compare that with the result of this call
You will need to use freezegun to make the PDF invariant and compare byte for byte

File: tests/test_pyfepdf.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        lineas_max=int(conf_fact.get("lineas_max", 24)),
+        qty_pos=conf_fact.get("cant_pos") or "izq",
+    )   
+
+    assert chk == False
+
+
+def test_generar_qr():
+    fepdf.CUIT = "30000000007"
+    assert fepdf.GenerarQR()
```
Comment (reingart) [inline tests/test_pyfepdf.py]: Pease check the result of GenerarQR method (an url), with a expected result
Example:
```
assert fepdf.GenerarQR() == "..."
````

File: tests/test_pyfepdf.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+
+def test_main_prueba():
+    sys.argv = []
+    sys.argv.append("--prueba")
+    sys.argv.append("--debug")
+    main()
+
+def test_main_cargar():
+    sys.argv = []
+    sys.argv.append("--prueba")
```
Comment (reingart) [inline tests/test_pyfepdf.py]: In this test, do not use `--prueba` (it will generate a new file)
You can add a sample expected factura.json to the test directory

File: tests/test_pyfepdf.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+    sys.argv.append("--cargar")
+    main()
```
Comment (reingart) [inline tests/test_pyfepdf.py]: After calling main, please do a sanity check s
Example:
```
assert pyfpdf.fepdf.factura ==  {...}
```
If there is no check, the test only assures that the code is executed and no serious error is raised (eg NameError)
It cannot warranty that the code works as expected

File: tests/test_pyfepdf.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+
+def test_main_cargar_json():
+    sys.argv = []
+    sys.argv.append("--prueba")
```
Comment (reingart) [inline tests/test_pyfepdf.py]: do not use `--prueba`, see previous comment

File: tests/test_pyfepdf.py [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+    sys.argv.append("--cargar")
+    sys.argv.append("--json")
+    sys.argv.append("--entrada")
+    sys.argv.append("datos/facturas.json")
+    main()
```
Comment (reingart) [inline tests/test_pyfepdf.py]: add some sanity check here, at least that the invoice json was loaded correctly (see previous comment)

File: tests/test_pyfepdf.py [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+
+def test_main_grabar():
+    sys.argv = []
+    sys.argv.append("--prueba")
+    sys.argv.append("--grabar")
+    sys.argv.append("--debug")
+    main()
```
Comment (reingart) [inline tests/test_pyfepdf.py:316]: Please check that the saved invoice, factura.txt is complete
You could compare it with a previous saved file

File: tests/test_pyfepdf.py [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+
+def test_main_grabar_json():
+    sys.argv = []
+    sys.argv.append("--prueba")
+    sys.argv.append("--grabar")
+    sys.argv.append("--json")
+    sys.argv.append("--debug")
+    main()
```
Comment (reingart) [inline tests/test_pyfepdf.py:334]: Idem to previous comment, compare the generated factura.json file with the one created by the test.

File: tests/test_pyfepdf.py [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+
+def test_mostrar_pdf():
+    sys.argv = []
+
+    config = SafeConfigParser()
+    config.read(CONFIG_FILE)
+    conf_fact = dict(config.items("FACTURA")) 
+
+    salida = conf_fact.get("salida", "")
+    fepdf.MostrarPDF(archivo=salida)
```
Comment (reingart) [inline tests/test_pyfepdf.py]: Not sure if thill will fail as evince is not installed or no display is available
You could use [pytest mock ](https://github.com/pytest-dev/pytest-mock/)to replace real os.system with a dummy function and check it was called

**Review by reingart — CHANGES_REQUESTED** [review]
Hi @chazuttu , the PR is very good!
I left some comments, please take a look and let me know if you need something to address them

File: .coveragerc (hunk 1 above)
Comment (chazuttu) [reply .coveragerc:9]: Issue created fir this as requested #90.

File: formatos/formato_json.py (hunk 2 above)
Comment (chazuttu) [reply formatos/formato_json.py:49]: @reingart this works well for both the python versions, without any error

File: formatos/formato_txt.py (hunk 3 above)
Comment (chazuttu) [reply formatos/formato_txt.py:262]: Issue Created --- #91 .

File: tests/test_pyfepdf.py (hunk 6 above)
Comment (chazuttu) [reply tests/test_pyfepdf.py:284]: Issue Created #92.

File: tests/test_pyfepdf.py (hunk 12 above)
Comment (chazuttu) [reply tests/test_pyfepdf.py:316]: Issue created #93.

**Review by reingart — APPROVED** [review]
